[JIT] Adding a concat optimization pass

navahgar opened the pull request:
This PR adds a new pass in JIT that optimizes `aten::cat` ops. 

Specifically, here are optimizations performed:
* Eliminate redundant in `cat` inputs by performing cse on the list of inputs.
   - This includes eliminating fully redundant `cat` ops when all the inputs are the same as well the case when "all but one" of the inputs have already been concatenated.
* Expand `cat` into multiple copies and eliminate redundancies.
   - This also includes eliminating redundancies in the underlying buffers used for `cat`.

These optimizations are not enabled in any compilation flow at this point.

navahgar committed:
[JIT] Adding a concat optimization pass

ZolotukhinM commented on torch/csrc/jit/passes/concat_opt.cpp:
```diff
@@ -0,0 +1,480 @@
+  }
+}
+
+class ConcatCseEliminator {
```
Nit: "CseEliminator" is a tautology I think

navahgar replied:
Right, will update the name.

navahgar replied:
Updated.

ZolotukhinM commented on torch/csrc/jit/passes/concat_opt.cpp:
```diff
@@ -0,0 +1,480 @@
+  void moveBefore(Node* node, Node* before) {
+    // In order to move a node before another node, we need to move
+    // all the nodes it depends on as well.
+    for (auto inp : node->inputs()) {
+      moveBefore(inp->node(), before);
+    }
```
This probably could be illegal if there is aliasing involved.

navahgar replied:
Can you give me an example?

ZolotukhinM reviewed (commented):
The code seems mostly fine, but I feel uneasy about this. It seems like we're pattern-matching a specific benchmark with this pass. I'd like to hear what other people think on this approach, @bertmaher, @Krovatkin, @eellison?

navahgar commented:
> The code seems mostly fine, but I feel uneasy about this. It seems like we're pattern-matching a specific benchmark with this pass. I'd like to hear what other people think on this approach, @bertmaher, @Krovatkin, @eellison?

The first optimization, `ConcatCse`, is looking for a rare pattern, which just happens to be in only one benchmark at this point. It also eliminates fully-redundant `cat` ops, which might be more common.

The second optimization, `ExpandCat`, applies to all `cat` ops, and there is nothing specific to any benchmark there. This is actually a precursor to an important optimization that will follow: using`out-` variants on the input ops.

penguinwu commented on test/cpp/jit/test_concat_opt.cpp:
```diff
@@ -0,0 +1,451 @@
+          # CHECK: prim::ListConstruct
+          %6 : Tensor[] = prim::ListConstruct(%4, %5)
+          # CHECK: aten::cat
+          %7 : Tensor = aten::cat(%6, %2)
```
Did not understand this example:
- Why do we need to know the shape of %7 for the concat optimization?
- Couldn't JIT automatically figure out the shape of %7 based on that of %6 and %2?

navahgar replied:
We can figure that out but that would be doing shape inference here. I'd prefer to keep that a separate pass and just handle `cat` optimizations here.

penguinwu replied:
Right. It makes sense to separate concerns on inferring shapes to a different pass.

penguinwu commented on test/cpp/jit/test_concat_opt.cpp:
```diff
@@ -0,0 +1,451 @@
+  checkOutputs(orig_outputs, opt_outputs);
+
+  // No optimizations should have happened in this case since the output
+  // shape of `aten::cat` is not known.
```
output shape of %7 is known:

```
%7 : Float(96, 56, 56, strides=[3136, 56, 1], requires_grad=0, device=cpu) = aten::cat(%6, %2)
```
Or did I miss something?

navahgar replied:
Oh, the comment is incorrect here. Here it does not get optimized because one of the inputs of `cat`, `%5`, does not have shape info. 

Sorry about that. Will fix the comment.

navahgar replied:
Fixed comment.

penguinwu commented on torch/csrc/jit/passes/tensorexpr_fuser.cpp:
```diff
@@ -1182,6 +1194,10 @@ void FuseTensorExprs(
   // Get rid of dead code so that we don't waste effort fusing it.
   EliminateDeadCode(graph);
 
+  if (catOptEnabled()) {
```
Is the concat optimization only beneficial when fuser is enabled or does it rely on anything from the fuser? 

If not, can we treat concat optimization as a general graph-level optimization?

navahgar replied:
The second optimization that expands `cat` into copies require shape info. That is the reason I am doing it in the fuser at this point. If we can get shape information by running a shape inference pass separately, then we can move this out of the fuser.

Technically, the `ConcatCse` opt can be done without shape information since that is only looking for common inputs. But that optimization relies on the input lists being materialized before every concat (and that the list operations have been removed which is taken care of by using a traced graph in my perf analysis).

penguinwu replied:
OK. Then it makes sense to keep it in the fuser for now.

penguinwu commented on torch/csrc/jit/passes/concat_opt.cpp:
```diff
@@ -0,0 +1,480 @@
+          prev_list_inputs.end(),
+          std::inserter(diff, diff.begin()));
+      if (diff.size() == 0) {
+        redundant_cats_[node] = it.second;
```
What if curr_list and prev_list contain the same elements but in different orders, e.g., `aten::cat(%1, %2, %3)` vs. `aten::cat(%2, %1, %3)`? Would `diff.size() == 0` in this case?

penguinwu replied:
Probably need to check if mutation happened to any of the elements in between the two concat lists. For instance,

```
%100 = aten::cat(%1, %2, %3)
aten::add_(%1, %1, ...) // in-place update on %1
%200 = aten::cat(%1, %2, %3)
```
In this case, %100 and %200 are not CSE

navahgar replied:
> What if curr_list and prev_list contain the same elements but in different orders, e.g., aten::cat(%1, %2, %3) vs. aten::cat(%2, %1, %3)? Would diff.size() == 0 in this case?

No. `diff` wouldn't be empty in this case. These would be treated as different since their concatenation order is different.

> Probably need to check if mutation happened to any of the elements in between the two concat lists. 

Good point. Added this check and a test for this. I wasn't doing this earlier since I was running it only on the traced graph. But this guard should be there as you pointed out.

Currently, the optimization will not happen for a `cat` op whenever the list that is input to that op has any mutations in the graph (only checking for `aten::append` for now). This could be improved by looking for mutations only in the region of optimization. That is a TODO for later.

penguinwu commented on torch/csrc/jit/passes/concat_opt.cpp:
```diff
@@ -0,0 +1,480 @@
+    }
+  }
+
+  void removeRedundantSlices() {
```
Do not quite understand what this does. Please add comments on what constitute as redundant slices (example would be great)

navahgar replied:
I had to comment explaining this above. It makes more sense here.. Moved it. PTAL.

penguinwu commented on torch/csrc/jit/passes/concat_opt.cpp:
```diff
@@ -0,0 +1,480 @@
+  ExpandConcatAndEliminateRedundancy(graph);
+  ConstantPooling(graph);
+  EliminateDeadCode(graph);
+  GRAPH_DUMP("After ConcatOpt", graph);
```
Do we need to update aliasdb after node deletion and generating slice and copy_ like ops @eellison ?

eellison replied:
AliasDb is instantiated for each pass, not maintained globally so no need to do any work here

penguinwu reviewed (changes requested):
Thanks for putting in a major feature like this. Those unit tests made my head spinning, not trivial patterns at all! I think I understood 80% of the codes. Please address the comments and I'll give it another thorough read again.

eellison commented on torch/csrc/jit/passes/tensorexpr_fuser.cpp:
```diff
@@ -1182,6 +1194,10 @@ void FuseTensorExprs(
+  if (catOptEnabled()) {
+    OptimizeConcat(graph);
+  }
+
```
I will get to a full review of this, but the types here aren't guaranteed, the contract is that the optimization has to set up it's own guards to do this... However it's possible we dont need shapes here..

navahgar replied:
I have added guards in the pass to ignore cases when shapes are unknown. We can probably do some of this with partial shapes. That is for a follow up though.

eellison replied:
the types present here aren’t “real” types, they are suggested types by profiling. They are not guaranteed to be correct. Cc @ZolotukhinM do u have the design doc of new profiling executor that explains this ?

navahgar replied:
So, you mean we need guards in the generated code in order to use these types that are coming out of the profiling executor? 

I'm assuming there is no other place in the compilation flow where we have even partial type info. Is this a valid assumption?

ZolotukhinM replied:
> Do u have the design doc of new profiling executor that explains this ?

No, we haven't documented this anywhere AFAIK :/

> So, you mean we need guards in the generated code in order to use these types that are coming out of the profiling executor?
> I'm assuming there is no other place in the compilation flow where we have even partial type info. Is this a valid assumption?

Types encoded in the IR at that point are types observed during profiling runs. They are not guaranteed to be the same on subsequent runs and hence optimization passes using these info are responsible for inserting appropriate runtime checks. In the fuser it is done here: https://github.com/pytorch/pytorch/blob/ba320cec6bc0e20855f8f6ce74f4248dcfafff5e/torch/csrc/jit/passes/tensorexpr_fuser.cpp#L1170-L1173

In the end the type info is cleared from the IR: https://github.com/pytorch/pytorch/blob/ba320cec6bc0e20855f8f6ce74f4248dcfafff5e/torch/csrc/jit/passes/tensorexpr_fuser.cpp#L581

This design kinda limits us at expressing types that are guaranteed to be the same across all runs - we discussed that in the past and @eellison had an idea of encoding profiled types as a node attribute rather than a type of a value. But we didn't do that for a reason I do not remember :)

eellison replied:
For the design doc, I meant the google doc u made

eellison replied:
I didn’t do it because we don’t have any facility in profiling executor to guarantee types, so there wasn’t much of a point in in implementing it.  It was discussed in relation to guarding autodiff graphs but I didn’t see much of a perf gain there for complexity so I didn’t do it.

ZolotukhinM replied:
The doc is here: https://docs.google.com/document/d/1pPHKgAk5qLKXWq4bSzosmY_wltcLd0ZE_CCclwsRmjQ/edit?ts=5f1a60ea#heading=h.vkwk1deu6xpt

But I think it doesn't reflect the actual design anymore, it describes the initial design relying on type maps, which we decided not to use.

navahgar replied:
I get the point now. It is clear that if we retain this pass in the current location in the compilation flow and use the types from profiling executor, we need guards in the code.

The alternative is to do this after the shape analysis that @eellison is working on, when we might have partial type info which might be good enough for this optimization. 

Any preference / suggestions?

ZolotukhinM replied:
> I didn’t do it because we don’t have any facility in profiling executor to guarantee types, so there wasn’t much of a point in in implementing it. It was discussed in relation to guarding autodiff graphs but I didn’t see much of a perf gain there for complexity so I didn’t do it.

I think it might be still worth reconsidering: the TorchScript IR could be used beyond the context of PE and it might be convenient to be able to express "guaranteed by something" vs "expected from the past runs" type info. A hypothetical scenario: we specialize the graph for given input sizes (and these sizes are checked somewhere outside the JIT runtime), but then we run it with profiling executor to refine some types of intermediate values. In this scenario we could've encoded types guaranteed externally and inferred from that as types of IR values and types observed during profiling phase as attributes.

navahgar replied:
In relation to @ZolotukhinM's point, where would the types inferred by shape analysis stored? How will that be differentiated from the type info added by the profiling executor? @eellison

eellison replied:
At this point I am mostly envisioning shape analysis for use in Freezing, where types are guaranteed, and as a way to prove & generate at runtime symbolic shapes in fusion/tensorexpr. 

I think it's important to invest in our IR and infrastructure that can be reused across different workflows with stuff like shape and dtype/device analysis. Once we have better facilities around propagating and proving information than we can reconsider what the execution model will look. 

I am focusing on optimizations in freezing because the ergonomics of writing passes is much better. I would like to prove out performance wins there first. In the short-midterm, I would recommend people writing optimization passes for inference to focus on that as well. That is where shapes and other information will first be added and guaranteed. 

The current executor was proposed with the idea that graph cleanup is not important, and that fusion is the only optimization that matters. While that may or may not be true, I think it's important to focus on particular optimizations we would like to enable before putting more work into the executor. And if the original thesis becomes less true, we might consider switching to a new execution model instead of incrementally adding features to the current one.

navahgar replied:
> I am focusing on optimizations in freezing because the ergonomics of writing passes is much better. I would like to prove out performance wins there first. In the short-midterm, I would recommend people writing optimization passes for inference to focus on that as well. That is where shapes and other information will first be added and guaranteed.

Sure, that sounds like a good plan. I'm going to wait for your shape analysis pass to land and target this as an optimization after freezing.

navahgar replied:
Removed calling this optimization from `tensorexpr_fuser`.

eellison reviewed (commented):
Will get to full review tomorrow... 

Re: pattern matching specific benchmark, I think it's always a question of complexity versus gain. The batchmm pass pretty much only fires for custom lstm/rnns backward, but that's a pretty common use case, it speeds it up a good amount, and we haven't had to do too much work to maintain it. I think part of the way we get to being a fast compiler is by incrementally improving specific models. However, I'm not sure about adding this without it being on by default, and I still need to look at the code..

eellison reviewed (commented):
We should not land this before confusion about the types present at fuser is cleared up. These types are not guaranteed to be correct. To use them each optimization must prove and guard all types it accesses. That is the current contract of this executor...

navahgar commented:
> We should not land this before confusion about the types present at fuser is cleared up. These types are not guaranteed to be correct. To use them each optimization must prove and guard all types it accesses. That is the current contract of this executor...

Now that we have decided to enable this as an optimization after freezing, I have removed it from the fuser workflow.

I would still like to land this pass and enable (and update if needed) once shape analysis is complete.

navahgar commented:
> Thanks for putting in a major feature like this. Those unit tests made my head spinning, not trivial patterns at all! I think I understood 80% of the codes. Please address the comments and I'll give it another thorough read again.

Addressed your comments. PTAL.

penguinwu reviewed (commented):
All my comments are addressed. Thanks Raghavan. Since others still have comments, I'll let others to approve the PR.

eellison commented on torch/csrc/jit/passes/concat_opt.cpp:
```diff
@@ -0,0 +1,465 @@
+    cat_out_size_list->insertBefore(node);
+
+    // Create an empty buffer to be used as `cat` output buffer.
+    auto cat_out_empty = graph_->create(
```
So, this is inserting the `aten::empty(int[] size, *, int? dtype, int? layout, Device? device, bool? pin_memory, MemoryFormat? memory_format=contiguous_format) -> Tensor`. schema and not specifying any value other than the size.

This only works for cpu float tensors as written because that is the default tensor instantiated with `empty`, but would break on any non-float non-cpu inputs.

navahgar replied:
What is the correct way to instantiate an `empty` tensor here?

eellison replied:
Ideally we'd have an operator that took in a Tensor and a size, and initialized a Tensor to that size. Until that exists you just have to manually insert nodes to get the `dtype/layout/device/memory_format` from one of the inputs. [Here](https://github.com/pytorch/pytorch/pull/52175/files#diff-c098125f9cb12f6523dba61c6eebd88ba0f49914261bc3879b83526c81b49991R45) is an example of doing that. 

This also begs another question: how does this optimization work with different memory_formats? What's the eager behavior on concat with different memory formats / non-contiguity ?

navahgar replied:
I'm inclining towards restricting this optimization to only contiguous Tensors.

eellison reviewed (commented):
Cool!

So addressing each optimization separately: 

First one looks good, I just had a few comments about using the AliasDb for correctness (please use lazily initialize style like https://github.com/pytorch/pytorch/pull/54640 although the returning bool for made change is not relevant). The nice thing about that the first optimization is that we don't need any types in order to do the transformation. I'm not opposed to model specific optimizations, but if we land code and maintain it we should be getting some juice out of it. I think that means turning on maybe as a freezing optimization or in inference. I think potentially for these types of less common inference optimizations it might make more sense to put in freezing where we have a bit more room for being slow.

It would be nice to get this to fire on a scripted densenet: 
I think the way to do this would be to add a JIT variant of `cat` that has a variadic number of inputs instead of a list. This sort of optimization would be useful outside of just getting this optimization to fire.  
Once we have that in place, the optimization on scripted models becomes a little easier.
you have:
```
li = [a]
out_1 = torch.cat(li)
li.append(b)
out_2 = torch.cat(li)
```
And then you iteratively try to move `li` to `cat`, or to `append`, so that you get.
```
out_1 = torch.cat(...a) # using variadic form
out_2 = torch.cat(...a, b)
``` 
And from there its easy to make the optimization you wrote kick in. We dont necessarily have to do it for this PR but it could be useful as a way to get densenet fully run in mkldnn once we add concat (cc @Krovatkin). By rewriting the concat with a variadic structure it helps avoid having to deal with mkldnn in containers (lists), which complicates the pass. 


For the second optimization: 

One difficulty here is that by rewriting the concat you are making it no longer fusable. Fusion is a more profitable optimization (especially on gpu) and we should prefer it over this optimization. I think this optimization is functionally equivalent to making a single node concat fusion group.  Additionally, as you said, the goal here is to make concat write to a reused memory buffer, however that would fall out naturally of reusing tensor buffers for outputs in fusion groups which is a more general feature. So I guess my opinion is we shouldnt land the second PR. 

Btw, next time you might consider doing a stacked PR with each optimization separately.

navahgar commented on torch/csrc/jit/passes/concat_opt.cpp:
```diff
@@ -0,0 +1,518 @@
+  }
+}
+
+bool isListMutatedInGraph(Value* list, const Graph& graph) {
```
That is good to know. Updated the pass to use `AliasDb`.

navahgar commented on torch/csrc/jit/passes/concat_opt.cpp:
```diff
@@ -0,0 +1,518 @@
+    //   %inputs = prim::ListConstruct(%inp1, %inp2, %inp3)
+    //   %concat = aten::cat(%inputs, %concat_dim)
+
+    // First, we check if %inputs is same as the inputs of any of the
```
I don't think it regular CSE is going to catch this case, does it? I believe this should be fairly straight-forward here AFAICT.

navahgar replied:
I somehow assumed CSE did not handle `ListConstruct` at all. Good to know that it is already covered. Removed that part from the code here.

navahgar commented on torch/csrc/jit/passes/concat_opt.cpp:
```diff
@@ -0,0 +1,518 @@
+      : graph_(std::move(graph)) {}
+
+  void run() {
+    for (auto node : graph_->block()->nodes()) {
```
Good point. Updated.

navahgar commented on torch/csrc/jit/passes/concat_opt.cpp:
```diff
@@ -0,0 +1,518 @@
+    //    %13 = aten::cat(%12, ...)
+    //    ...
+    //        = %13 ... // Use %13
+    auto curr_inputs_prefix = curr_inputs.slice(0, curr_inputs.size() - 1);
```
That's a good idea. But yeah, lets do that when we see some real cases for that.

navahgar commented on torch/csrc/jit/passes/concat_opt.cpp:
```diff
@@ -0,0 +1,518 @@
+      if (equal(curr_inputs_prefix, it.first->inputs())) {
+        std::vector<Value*> new_list_values = {
+            it.second->output(), curr_inputs.back()};
+        auto new_list_node =
```
Done.

navahgar commented on torch/csrc/jit/passes/concat_opt.cpp:
```diff
@@ -0,0 +1,518 @@
+    graph_->setInsertPoint(node);
+    auto none = graph_->insertConstant(IValue());
+    auto one = graph_->insertConstant(1);
+    auto empty_symbol = Symbol::aten("empty");
```
They are already interned. Replaced these to use that.

navahgar commented on torch/csrc/jit/passes/concat_opt.cpp:
```diff
@@ -0,0 +1,518 @@
+    auto copy_symbol = Symbol::aten("copy_");
+
+    // Insert the constants needed for the `cat` output buffer size.
+    auto tensortype = dynamic_cast<TensorType*>(node->output()->type().get());
```
Good to know. Updated.

navahgar commented on torch/csrc/jit/passes/concat_opt.cpp:
```diff
@@ -0,0 +1,465 @@
+    int cat_dim_value = cat_dim->node()->i(attr::value);
+
+    // Set the insertion point to the curent `cat` node.
+    graph_->setInsertPoint(node);
```
That is very useful. Thanks for pointing that out. Updated.

navahgar commented:
@eellison Thanks a lot for the detailed review. Addressed your comments. PTAL.

Regarding your high-level comments on the optimizations:

1. Your idea to get it fire on scripted densenet is interesting. I want to talk more about this offline with you. 

2. While this optimization makes `cat` ops non-fusable, the idea is to combine this with using the out-variant on inputs to `cat` (which is something I am working on as a follow up). That would essentially eliminate `cat` completely. Why would that not be useful? Also, with `cat` in fusion groups, the buffers do not get reused (at least not in NNC).

eellison reviewed (commented):
Commented about empty, getting to rest of re-review first thing tomorrow

eellison commented on torch/csrc/jit/passes/concat_opt.cpp:
```diff
@@ -0,0 +1,487 @@
+    if (getOrCreateAliasDb()->hasWriters(node->input(0))) {
+      return;
+    }
+    if (!allShapesAreKnown(node)) {
```
We might have to think a bit about how this works with non-contiguous Tensors. For one, if the inputs are all channels last, then the output should be too. I think other than that this is fine ? slice_ should handle if the input to it is non-contiguous.

navahgar replied:
I agree. I haven't given much thought about handling non-contiguous Tensors. I think it would be better if we restrict this optimization to contiguous tensors. 

How can I check if the Tensors involved in this cat are contiguous?

eellison replied:
Let’s leave this as follow up / TODO

eellison commented on torch/csrc/jit/passes/concat_opt.cpp:
```diff
@@ -0,0 +1,487 @@
+      slices_added_.push_back(slice);
+
+      // Insert a copy from this input to the output slice.
+      auto copy = graph_->create(copy_symbol, {slice->output(), cat_inp});
```
nit: I think it would be more readable if `copy_symbol` was inline. as is, i had to look back to double check that this was `copy_` and not `copy`. (the second one doesnt exist, but i didnt know that off hand reading the code)

navahgar replied:
Updated.

eellison commented on torch/csrc/jit/passes/concat_opt.cpp:
```diff
@@ -0,0 +1,487 @@
+    }
+
+    // Replace the uses of `cat` node with the cat output buffer.
+    replace_uses_with_[node->output()] = cat_out_value;
```
Nice, yea i was going to comment about this. You need to defer the actuall modification of the graph until after doing the analysis or you'll invalidate the AliasDb (meaning yea you are doing it correctly right now)

eellison commented on torch/csrc/jit/passes/concat_opt.cpp:
```diff
@@ -0,0 +1,487 @@
+    node->moveBefore(before);
+  }
+
+  // Reuse buffers in copies wherever possible.
```
Hmm, I'm having a little bit of a difficult time understanding this one. copy_ has broadcasting behavior, we might have to be a little careful. Do you mind rewriting the example you have here as python code / explaining a bit more ?

navahgar replied:
Added some more explanation for this. PTAL.

navahgar commented on torch/csrc/jit/passes/concat_opt.h:
```diff
@@ -0,0 +1,16 @@
+TORCH_API void ExpandConcatAndEliminateRedundancy(
+    const std::shared_ptr<Graph>& graph);
+
+TORCH_API void OptimizeConcat(const std::shared_ptr<Graph>& graph);
```
Added.

navahgar commented on torch/csrc/jit/passes/concat_opt.cpp:
```diff
@@ -0,0 +1,487 @@
+
+    auto curr_list = node->input(0)->node();
+    auto curr_inputs = curr_list->inputs();
+    for (const auto& it : concated_lists_) {
```
Done.

navahgar commented on torch/csrc/jit/passes/concat_opt.cpp:
```diff
@@ -0,0 +1,487 @@
+      // Can't expand when shapes are not known for the `cat` op.
+      return;
+    }
+    for (auto cat_inp : node->input(0)->node()->inputs()) {
```
I'm already checking if `node->input(0)` has writers above. Added check for the latter.

navahgar commented on torch/csrc/jit/passes/concat_opt.cpp:
```diff
@@ -0,0 +1,487 @@
+        return;
+      }
+    }
+    auto cat_dim = node->input(1);
```
Thanks for pointing that out. That's useful to know. Updated.

navahgar commented on torch/csrc/jit/passes/concat_opt.cpp:
```diff
@@ -0,0 +1,487 @@
+    // Do not optimize cat nodes whose inputs are mutated in the graph.
+    // TODO: Improve this by checking if it is mutated in the graph region
+    // where this optimization is applied.
+    if (getOrCreateAliasDb()->hasWriters(node->input(0))) {
```
Good point. I have added a check to not consider any cat node whose output has writes for future optimizations. Note that that particular cat (which has a write) could itself be optimized using some previous cat. That is okay since the output of this is never used for this optimizations.

eellison reviewed (approved):
LGTM! Thanks for addressing feedback! We should definitely take it as a follow up to get these firing on scripted /frozen models once shape analysis lands

navahgar commented:
> LGTM! Thanks for addressing feedback! We should definitely take it as a follow up to get these firing on scripted /frozen models once shape analysis lands

Thanks for your detailed reviews. They were very helpful.

penguinwu reviewed (approved):
LGTM
